### PR TITLE
shrank the binary and image size

### LINKED ISSUE #352: MAINTAINER instruction in Dockerfile is deprecated
MAINTAINER should be replaced by LABEL maintainer="..."
See https://docs.docker.com/engine/reference/builder/#maintainer-deprecated

### PR DESCRIPTION
**Pull request by nrvnrvn**
Changes:
- added a new ldflag to make the resulting binary a static ELF;
- removed debug symbols to make it smaller;
- switched to using `SCRATCH` as the base image;
- fixed #352 by coincidence.

### CODE REVIEW

File: .promu.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ build:
         -X github.com/prometheus/common/version.Branch={{.Branch}}
         -X github.com/prometheus/common/version.BuildUser={{user}}@{{host}}
         -X github.com/prometheus/common/version.BuildDate={{date "20060102-15:04:05"}}
+        -s -w -extldflags "-static"
```
Comment (SuperQ) [inline .promu.yml]: This should probably match the upstream [promu config](https://github.com/prometheus/promu/blob/master/.promu.yml#L4-L6).

**Comment by SuperQ** [comment]
I think we do want to keep the debug symbols. Also, I don't think switching to `FROM scratch` is appropriate to do here without a discussion on prometheus-developers@googlegroups.com first. We like to keep our various projects consistent, and this would be a change we would want to have the same everywhere.

**Comment by SuperQ** [comment]
I'm pretty sure this is unecessary, our build default produces a static binary.

```
$ file blackbox_exporter
blackbox_exporter: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
```

File: .promu.yml (hunk 1 above)
Comment (nrvnrvn) [reply .promu.yml]: yep, you were right about the default static. The config now matches the upstream.

**Comment by SuperQ** [comment]
I don't see why this is necessary. This seems to eliminate our busybox image from the image, which means the usual busybox tools will not be available in the image.

**Comment by nrvnrvn** [comment]
The reason is to keep the image as small as possible. When it goes for static linux binaries it is always tempting to use scratch as the base image. See [coredns](https://github.com/coredns/coredns/blob/master/Dockerfile) for instance. If for whatever reason you need to debug you can always inject busybox inside using `kubectl cp`/`docker cp`.

I hadn't looked into other Dockerfiles under https://github.com/prometheus until you mentioned consistency. This is a good point. I would not be happy to bring inconsistency to the official prometheus images. Now I see that most of your images use `https://github.com/prometheus/busybox` as the base image.

Feel free to close this PR and if you think it is worth raising the question of switching from `quay.io/prometheus/busybox:latest` to `scratch` wherever possible in prometheus-developers@googlegroups.com please let me know.

**Comment by brian-brazil** [comment]
What's the size difference here?

**Comment by SuperQ** [comment]
The uClibc busybox image is 1.46MiB.

**Comment by SuperQ** [comment]
Yes, we should have a discussion first.

We already have an extremely compact busybox image. I'm a fan of `from scratch` containers, especially for Go. But I see the utility of having some basics in the container for inspection.

We could always propose having a `debug` type tag on Docker that contains the busybox image, but default to a scratch image.

Looking forward to the discussion.

**Pull request closed without merging**